<html>
<head>
    <!--plotly scripts -->
    <!-- <script src="https://cdn.plot.ly/plotly-latest.min.js"></script> -->
<script src="{{ url_for('static', filename='js/third-party/plotly/dist/plotly.min.js') }}"></script>
    <!--bootstrap scripts -->
   <!-- <link href= "{{ url_for('static', filename='css/bootstrap.min.css')}}" rel="stylesheet">
    <script src="//code.jquery.com/jquery-1.11.1.min.js"></script>
    <script type="text/javascript" src="{{ url_for('static', filename='js/bootstrap.min.js') }}"></script>-->
	    <link href= "{{ url_for('static',
	filename='js/third-party/bootstrap/dist/css/bootstrap.min.css')}}" rel="stylesheet">
    <script src="{{ url_for('static', filename='js/third-party/jquery/dist/jquery.min.js') }}"></script>
	<script type="text/javascript" src="{{ url_for('static', filename='js/third-party/bootstrap/dist/js/bootstrap.min.js') }}">
	</script>


</head>
<body>
<h2> K-means Clustering </h2> <!--title goes here-->
<p>   </p>

<div id="controls">
<form class="form inline" action="https://geco.iplantcollaborative.org/hclent/bionlp/cogekmeans/" method="post">
<label for="buttongroup">clusters:</label>
	<div id="buttongroup">
		<div class="col-xs-1">
		<select class="form-control" id="k_val" name="k_val">
		  <option name="1" value="1">k=?</option>
	  <option name="2" value="2">k=2</option>
	  <option name="3" value="3">k=3</option>
	  <option name="4" value="4">k=4</option>
	  <option name="5" value="5">k=5</option>
		   </select>
		</div>

		<button type="submit" id="kmeans" class="btn btn-default" data-toggle="modal" data-target="#myModal">Update</button>
		</div>
</form>
</div>

<!-- Progress modal -->
<div class="modal fade" id="myModal" tabindex="-1" role="dialog" aria-labelledby="myModalLabel" aria-hidden="true">
  <div class="modal-dialog" role="document">
    <div class="modal-content">
      <div class="modal-header">
        <button type="button" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
        <h4 class="modal-title" id="myModalLabel">Loading</h4>
      </div>
      <div class="modal-body">

				<!-- Progress bar that doesn't actually move -->
				<div class="progress">
				  <div class="progress-bar progress-bar-success progress-bar-striped active" role="progressbar" aria-valuenow="1" aria-valuemin="0" aria-valuemax="100" style="width: 45%">
				    <span class="sr-only">45% Complete</span>
				  </div>
				</div>

      </div>
      <div class="modal-footer">
        <button type="button" class="btn btn-secondary" data-dismiss="modal">Close</button>
      </div>
    </div>
  </div>
</div>

<br></br>
<br></br>


<div id="myDiv">  </div>
<div id="clickInfo"> </div>

<script>
var myPlot = document.getElementById('myDiv'),
	clickInfo = document.getElementById('clickInfo'),

	trace1 = {
	x: [0.036005355418496574, 0.10899050539529827, 0.033117369063888444, 0.12801409554567816, 0.10924823930772593, 0.12278301511465704, 0.1279917038417461, 0.033644503526268937, 0.0, 0.092801391602944344, 0.032003664571078484, 0.037104472535857844, 0.034094719373483837, 0.01831241991030096, 0.075824279720484056, 0.04407615007610579, 0.076164215689938239, 0.12251498618527518, 0.044597595057650159, 0.036005355418496574, 0.0035444053102094654],
	y: [0.0, 0.045353993890903249, 0.0, 0.017343475861352896, 0.0, 0.0, 0.0023355496821541341, 0.027523532125564597, 0.11195046137655965, 0.0, 0.0029180793861159743, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.035063603719054447],
	z: [0.010306260295888653, 0.037473716871444344, 0.14155982751483784, 0.1248359598374285, 0.040609076988801833, 0.095857331241418042, 0.15552190452509324, 0.094493507013691719, 0.035879142314698576, 0.035657308700375284, 0.022951367193510302, 0.045292199157505098, 0.074530070841575438, 0.062806245741434827, 0.029628570489764573, 0.021253118888752056, 0.15178540446729008, 0.18987788705125369, 0.038029270827085231, 0.010306260295888653, 0.17818747436617241],
	name: ['A Genomic Analysis of Factors Driving lincRNA Diversification: Lessons from Plants', 'The N-Terminus of the Floral Arabidopsis TGA Transcription Factor PERIANTHIA Mediates Redox-Sensitive DNA-Binding', 'The genome sequence of the outbreeding globe artichoke constructed de novo incorporating a phase-aware low-pass sequencing strategy of F1 progeny', 'Detecting and Locating Whole Genome Duplications on a Phylogeny: A Probabilistic Approach', 'Whole genome duplication events in plant evolution reconstructed and predicted using myosin motor proteins', 'Synonymous codon usage bias is correlative to intron number and shows disequilibrium among exons in plants', 'The tomato genome: implications for plant breeding, genomics and evolution', 'The iPlant Collaborative: Cyberinfrastructure for Plant Biology', 'Paleopolyploidy in the Brassicales: Analyses of the Cleome Transcriptome Elucidate the History of Genome Duplications in Arabidopsis and Other Brassicales', 'Dosage Sensitivity of RPL9 and Concerted Evolution of Ribosomal Protein Genes in Plants', 'Comparative Genome Analyses Reveal Distinct Structure in the Saltwater Crocodile MHC', 'Gene families as soft cliques with backbones: Amborella contrasted with other flowering plants', 'The dynamics of functional classes of plant genes in rediploidized ancient polyploids', 'Ancient eudicot hexaploidy meets ancestral eurosid gene order', 'Fractionation, rearrangement and subgenome dominance', 'Gene order in rosid phylogeny, inferred from pairwise syntenies among extant genomes', 'Altered Patterns of Fractionation and Exon Deletions in Brassica rapa Support a Two-Step Model of Paleohexaploidy', 'Structure and evolution of the plant cation diffusion facilitator family of ion transporters', 'Bovipain-2, the falcipain-2 ortholog, is expressed in intraerythrocytic stages of the tick-transmitted hemoparasite Babesia bovis', 'Partitioning the Apical Domain of the Arabidopsis Embryo Requires the BOBBER1 NudC Domain Protein', 'Finding and Comparing Syntenic Regions among Arabidopsis and the Outgroups Papaya, Poplar, and Grape: CoGe with Rosids'],
	mode: 'markers',
	marker: {
		color: 'rgb(110, 220, 110)',
		size: 12,
		symbol: 'circle',
		line: {
		color: 'rgb(110, 220, 110)',
		width: 0.5},
		opacity: 0.8},
	type: 'scatter3d'
	};

	trace2 = {
	x: [0.2144063014032179, 0.20545435832718226, 0.16465230943768017, 0.28935612214028356, 0.19512256894080879, 0.22490261256128133, 0.23473015749611748, 0.22711316431654685, 0.22956957377371939, 0.27483075569770071, 0.19690525302066525, 0.26592577158014818, 0.12727577855181382, 0.25348582532471653, 0.22225033520269541, 0.22605367936649748, 0.26843144054580287, 0.29696648827528516, 0.17421521364522602, 0.23872047024554663, 0.32756352750454154, 0.13365794560477337, 0.20465833294325672, 0.29637080770926094, 0.28777028823234951, 0.26382889166758622, 0.26150122222675559, 0.2519086706136926, 0.27786509448700147, 0.26652180486018506, 0.27941069178516864, 0.23357443458007465, 0.16340787608922061, 0.1530003766871269, 0.20368043086834944, 0.2517038079726816, 0.21906287419590359, 0.13508638281842084, 0.12623563780243416, 0.15293249931590563, 0.16204248487491368, 0.22490261256128133, 0.18645429709635755, 0.29637080770926094, 0.13365794560477337, 0.27483075569770071, 0.21906287419590359, 0.22373601314362199, 0.18914753568360126, 0.2144063014032179, 0.1960677115633738, 0.20860621729993981, 0.17595813906562072, 0.25641300540896533, 0.20465833294325672, 0.23872047024554663, 0.19690525302066525, 0.23410471791705267, 0.16619792778137807, 0.26382889166758622, 0.17778457846576803, 0.18403061524191128],
	y: [0.11173267351255603, 0.20433794110860146, 0.073265304677427631, 0.0, 0.01459835102449079, 0.14871909034048555, 0.082910425249419689, 0.0, 0.0, 0.0, 0.047107277565278438, 0.0, 0.082076284103451233, 0.0, 0.076639994679162354, 0.061672973769551276, 0.0, 0.0, 0.078904930971740381, 0.013541229998576534, 0.01767141223411128, 0.0, 0.10932325181477501, 0.025054581782691786, 0.0099313896634732023, 0.08533636395392849, 0.13377616036066631, 0.066946409340712859, 0.00023773813952989196, 0.0, 0.00068028728506405473, 0.0, 0.019992274913285302, 0.20269348348715144, 0.16690396091347767, 0.1147415979783092, 0.015888937103444151, 0.0, 0.055572682173820658, 0.038189612917747154, 0.028572801131696437, 0.14871909034048555, 0.0, 0.025054581782691786, 0.0, 0.0, 0.015888937103444151, 0.0731987158289958, 0.040228983406310363, 0.11173267351255603, 0.0, 0.0, 0.0, 0.0, 0.10932325181477501, 0.013541229998576534, 0.047107277565278438, 0.0, 0.038572774807105817, 0.08533636395392849, 0.0, 0.10991074873044017],
	z: [0.0, 0.001068410654863661, 0.1493322735231275, 0.055340594977346273, 0.051729012107484616, 0.040299699901106691, 0.0088222129018390612, 0.0, 0.093304799504715083, 0.0, 0.0, 0.050821538053954361, 0.018829721380324706, 0.058155959167489155, 0.080011826586482621, 0.033242393684595528, 0.050715275914372034, 0.02571283398248006, 0.0098041617525336921, 0.041612422126771555, 0.028876284120630716, 0.14303032321171863, 0.080733951587266109, 0.04004352157795172, 0.067074986371575177, 0.085201501490872827, 0.0024413100595223882, 0.078672560241149367, 0.052418605218583221, 0.071592442833580405, 0.034653754641279233, 0.055636106461927272, 0.11293999234759512, 0.028404632677743909, 0.025822151404715075, 0.020512223774748862, 0.0, 0.16299623798724061, 0.016866241364938781, 0.026179019578519813, 0.11149219473877132, 0.040299699901106691, 0.0, 0.04004352157795172, 0.14303032321171863, 0.0, 0.0, 0.039999965557457932, 0.099338002782851403, 0.0, 0.10890124694246769, 0.017487049039203128, 0.1175240858250773, 0.052658764413223265, 0.080733951587266109, 0.041612422126771555, 0.0, 0.063515550175722052, 0.10482614569988616, 0.085201501490872827, 0.083729024273943659, 0.13437656446981008],
	name: ['Evolution of Gene Duplication in Plants', 'Integration of Lupinus angustifolius L. (narrow-leafed lupin) genome maps and comparative mapping within legumes', 'Locating rearrangement events in a phylogeny based on highly fragmented assemblies', 'The pineapple genome and the evolution of CAM photosynthesis', 'Development and Genetic Control of Plant Architecture and Biomass in the Panicoid Grass, Setaria', 'VGSC: A Web-Based Vector Graph Toolkit of Genome Synteny and Collinearity', 'Visualizing genome and systems biology: technologies, tools, implementation techniques and trends, past, present and future', 'A comprehensive survey of the grapevine VQ gene family and its transcriptional correlation with WRKY proteins', 'Large-Scale Evolutionary Analysis of Genes and Supergene Clusters from Terpenoid Modular Pathways Provides Insights into Metabolic Diversification in Flowering Plants', 'Structure, expression profile and phylogenetic inference of chalcone isomerase-like genes from the narrow-leafed lupin (Lupinus angustifolius L.) genome', 'The plant Polycomb repressive complex 1 (PRC1) existed in the ancestor of seed plants and has a complex duplication history', 'PLAZA 3.0: an access point for plant comparative genomics', 'Lineage-specific evolution of Methylthioalkylmalate synthases (MAMs) involved in glucosinolates biosynthesis', 'CenH3 evolution in diploids and polyploids of three angiosperm genera', 'Regulation of cell divisions and differentiation by MALE STERILITY32 is required for anther development in maize', 'Polyploid Evolution of the Brassicaceae during the Cenozoic Era', 'Comparative Phylogenomics Uncovers the Impact of Symbiotic Associations on Host Genome Evolution', 'Polyploidy-associated genome modifications during land plant evolution', 'Distinct evolutionary strategies in the GGPPS family from plants', 'Genome-Wide Identification, Evolution and Expression Analysis of the Grape (Vitis vinifera L.) Zinc Finger-Homeodomain Gene Family', 'A near complete snapshot of the Zea mays seedling transcriptome revealed from ultra-deep sequencing', 'Cross-Family Translational Genomics of Abiotic Stress-Responsive Genes between Arabidopsis and Medicago truncatula', 'Description of durum wheat linkage map and comparative sequence analysis of wheat mapped DArT markers with rice and Brachypodium genomes', 'Practical aliquoting of flowering plant genomes', 'Locally Duplicated Ohnologs Evolve Faster Than Nonlocally Duplicated Ohnologs in Arabidopsis and Rice', 'Conservation of two lineages of peroxisomal (Type I) 3-ketoacyl-CoA thiolases in land plants, specialization of the genes in Brassicaceae, and characterization of their expression in Arabidopsis thaliana', 'Genome-Wide Identification and Analysis of the TIFY Gene Family in Grape', 'Preferential Retention of Circadian Clock Genes during Diploidization following Whole Genome Triplication in Brassica rapa', 'Order in the playground: Formation of plant gene clusters in dynamic chromosomal regions', 'Genome-Wide Identification and Analysis of Grape Aldehyde Dehydrogenase (ALDH) Gene Superfamily', 'Different Gene Families in Arabidopsis thaliana Transposed in Different Epochs and at Different Frequencies throughout the Rosids', 'Selection for Higher Gene Copy Number after Different Types of Plant Gene Duplications', 'Changes in Twelve Homoeologous Genomic Regions in Soybean following Three Rounds of Polyploidy', 'Dating and functional characterization of duplicated genes in the apple (Malus domestica Borkh.) by analyzing EST data', 'Genes Identified by Visible Mutant Phenotypes Show Increased Bias toward One of Two Subgenomes of Maize', 'An EST-based analysis identifies new genes and reveals distinctive gene expression features of Coffea arabica and Coffea canephora', 'Ancient genome duplications during the evolution of kiwifruit (Actinidia) and related Ericales', 'Genomicus: a database and a browser to study gene synteny in modern and ancestral genomes', 'Many or most genes in Arabidopsis transposed after the origin of the order Brassicales', 'MultiSyn: A Webtool for Multiple Synteny Detection and Visualization of User’s Sequence of Interest Compared to Public Plant Species', 'Locating rearrangement events in a phylogeny based on highly fragmented assemblies', 'SynFind: Compiling Syntenic Regions across Any Set of Genomes on Demand', 'Syntenic block overlap multiplicities with a panel of reference genomes provide a signature of ancient polyploidization events', 'Widespread mitovirus sequences in plant genomes', 'CenH3 evolution in diploids and polyploids of three angiosperm genera', 'A novel approach for multi-domain and multi-gene family identification provides insights into evolutionary dynamics of disease resistance genes in core eudicot plants', 'Regulation of cell divisions and differentiation by MALE STERILITY32 is required for anther development in maize', 'Genomic Characterization of the LEED..PEEDs, a Gene Family Unique to the Medicago Lineage', 'Genome and transcriptome sequencing identifies breeding targets in the orphan crop tef (Eragrostis tef)', 'Phylogenetic distribution of three pathways for propionate production within the human gut microbiota', 'HvLUX1 is a candidate gene underlying the early maturity 10 locus in barley: phylogeny, diversity, and interactions with the circadian clock and photoperiodic pathways', 'Whole Genome and Tandem Duplicate Retention Facilitated Glucosinolate Pathway Diversification in the Mustard Family', 'Comprehensive analysis of imprinted genes in maize reveals allelic variation for imprinting and limited conservation with other species', 'The Deacetylase Sir2 from the Yeast Clavispora lusitaniae Lacks the Evolutionarily Conserved Capacity to Generate Subtelomeric Heterochromatin', 'The Fragaria vesca Homolog of SUPPRESSOR OF OVEREXPRESSION OF CONSTANS1 Represses Flowering and Promotes Vegetative Growth', 'Automated conserved non-coding sequence (CNS) discovery reveals differences in gene content and promoter evolution among grasses', 'Comparative genomics of Lupinus angustifolius gene-rich regions: BAC library exploration, genetic mapping and cytogenetics', 'Genetic Control and Comparative Genomic Analysis of Flowering Time in Setaria (Poaceae)', 'PGDD: a database of gene and genome duplication in plants', 'Preferential Retention of Circadian Clock Genes during Diploidization following Whole Genome Triplication in Brassica rapa', 'Using Genomic Sequencing for Classical Genetics in E. coli K12', 'Proper regulation of a sperm-specific cis-nat-siRNA is essential for double fertilization in Arabidopsis'],
	mode: 'markers',
	marker: {
		color: 'rgb(255, 185, 51)',
		size: 12,
		symbol: 'circle',
		line: {
		color: 'rgb(255, 185, 51)',
		width: 1},
		opacity: 0.8},
	type: 'scatter3d'};

	trace3 = {
	x: [0.0, 0.0, 0.016721872629465454, 0.0, 0.0, 0.0, 0.0],
	y: [0.32303156306950909, 0.19982834824357362, 0.13218277240363044, 0.20123801858887938, 0.27467973521584083, 0.11829573625269293, 0.34750142852034294],
	z: [0.042729026188386005, 0.15120031503181325, 0.11728569233672072, 0.12772468044001592, 0.10885100985492834, 0.1795160424730084, 0.039819255666657352],
	name: ['Natural Variation in Sensitivity to a Loss of Chloroplast Translation in Arabidopsis',
		'A physical map for the Amborella trichopoda genome sheds light on the evolution of angiosperm genome structure',
		'The Evolution of a High Copy Gene Array in Arabidopsis',
		'Two Bacterial Genera, Sodalis and Rickettsia, Associated with the Seal Louse Proechinophthirus fluctus (Phthiraptera: Anoplura)',
		'Identification of dopamine receptors across the extant avian family tree and analysis with other clades uncovers a polyploid expansion among vertebrates',
		'Positionally-conserved but sequence-diverged: identification of long non-coding RNAs in the Brassicaceae and Cleomaceae',
		'Evolutionary and Experimental Assessment of Novel Markers for Detection of Xanthomonas euvesicatoria in Plant Samples'],
	mode: 'markers',
	marker: {
		color: 'rgb(49, 113, 231)',
		size: 12,
		symbol: 'circle',
		line: {
		color: 'rgb(49, 113, 231)',
		width: 1},
		opacity: 0.8},
	type: 'scatter3d'};


	 data = [trace1, trace2, trace3];
	 layout = {margin: {
	l: 0,
	r: 0,
	b: 0,
	t: 0 }};

Plotly.plot('myDiv', data, layout);


myPlot.on('plotly_click', function(data){
	for (var i=0; i<data.points.length; i++) { //safety net
		var selected_point = data.points[i];
		var point_number = data.points[i].pointNumber;
		var point_data = selected_point['data'];
		var point_title = point_data['name'][point_number];
		//var point_title = data.points[i]['data']['name'][data.points[i].pointNumber];
		//console.log(point_title);

		clickInfo.innerHTML = point_title;
	}
});  //end plotly_click



	</script>


<script>
  var btn = document.getElementById('kmeans');
  btn.addEventListener('click', loadProgress);

  function loadProgress() {
    //console.log('Button clicked!');
    var progressBar = document.getElementsByClassName('progress-bar')[0];
    var interval = 30;
    var currentValue = 0;
    for (var i = 0; i < 100; i += interval) {
      //console.log('Setting interval', i)
      setTimeout(function() {
        currentValue += i;
        //console.log('Setting progress width to:', currentValue, '%');
        progressBar.setAttribute('style', 'width: ' + currentValue + '%');
      }, i);
    }
  }
</script>